Doc(Migration): Add Helm v2 to v3 migration doc

hickeyma opened the pull request:
**What this PR does / why we need it**:
Adds user documentation for migration from Helm v2 to v3.

Partial-Closes: #5892

**Special notes for your reviewer**:

**If applicable**:
- [x] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

hickeyma committed:
Add v2 to v3 migration docs

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

bacongobbler commented:
Would you mind taking a look at the existing FAQ and see if this applies? https://github.com/helm/helm/blob/dev-v3/docs/faq.md#changes-since-helm-2

hickeyma commented:
> Would you mind taking a look at the existing FAQ and see if this applies? 

@bacongobbler Sure, I will do. Thanks.

hickeyma committed:
Update with latest changes to v3

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Have a few questions wrt v3:

- Will v3 have configuration and if so will it be the same directory as v2? Will still need to do a `helm init`?
- Is this correct: v2 uses "ConfigMaps" under Tiller namespace and ownership. v3 uses "Secrets" in user namespace and Helm ownership.
- Is the release object in v3:
    - Secret resource object
    - Incremental like v2
    - Release data a base-64 encoded, gzipped archive of the entire release record (like v2)

hickeyma commented:
@bacongobbler FYI, there is very little in the FAQ at the moment.

bacongobbler commented:
Yes. I've been working on the blog post about changes from Helm 2 to Helm 3. I was planning on porting that over to the docs. See part 1 [here](https://helm.sh/blog/helm-3-preview-pt1/)

bacongobbler commented:
Hey @hickeyma, now that #5718 has been merged, is there things inside this PR that should be added to that doc or is this safe to close?

hickeyma commented:
Hey @bacongobbler. I think this doc is a standalone doc that will be used by users when migrating to v3. #5718 provides a description of the changes in v3. This doc provides the steps on how to migrate to v3.

hickeyma committed:
Add more details to the migration doc

- Steps to mape v2 to v3 stste
- Update current v3 implementation details
- Add steps to map release to another namespace

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Update following changes in v3

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Opening this for review now. This doc will probably develop into better user documentation but lets start it from here.

rimusz commented on docs/v2_v3_migration.md:
```diff
@@ -11,58 +11,93 @@ Migration from v2 to v3 Helm
+   - The choices are as follows:
+     - Lose the release history: Deploy as new using v3 and delete the existing release using v2
+     - Maintain the release history: TBD: Detail the steps involved in migration of v2 history to v3. This should describe the steps manually and then any tools provided for those manual steps. Are the steps involved:
+       - Retrieve ConfigMaps for Tiller owner
```
v2 releases could also be stored in secrets

hickeyma replied:
Good point, thanks

hickeyma replied:
Sure. Updated

hickeyma committed:
Update following changes in beta and feedback

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Updated following beta changes and community feedback. Ready for review again.

php-coder commented on docs/v2_v3_migration.md:
```diff
@@ -0,0 +1,81 @@
+     - Authentication supported for repo login/logout (`helm registry <sub_command>`)
+     - Authorization (OAuth2) supported which enables chart upload/install access rights
+3. Chart apiVersion bumped to "v2" for following specification changes:
+   - Dynamically linked chart deopendencies moved to `Chart.yaml` (`requirements.yaml` removed and  requirements --> dependencies)
```
```suggestion
   - Dynamically linked chart dependencies moved to `Chart.yaml` (`requirements.yaml` removed and  requirements --> dependencies)
```

hickeyma replied:
done

php-coder commented on docs/v2_v3_migration.md:
```diff
@@ -0,0 +1,81 @@
+ 
+2. Migrating Helm v2 to Helm v3:
+   - Migrating Helm v2 configuration (`helm home`) to Helm v3:
+     - - Proposal is a standalone migration tool that migrates from Helm v2 to Helm v3: https://github.com/helm/helm/issues/6154
```
```suggestion
     - Proposal is a standalone migration tool that migrates from Helm v2 to Helm v3: https://github.com/helm/helm/issues/6154
```

hickeyma replied:
Sentence no longer applies.

php-coder commented on docs/v2_v3_migration.md:
```diff
@@ -0,0 +1,81 @@
+         - push: push a chart to remote
+         - remove: remove a chart
+         - save: save a chart directory  
+       - delete --> uninstall : removes all release history by default (previous;y needed `--purge`)
```
```suggestion
       - delete --> uninstall : removes all release history by default (previously needed `--purge`)
```

hickeyma replied:
done

php-coder commented on docs/v2_v3_migration.md:
```diff
@@ -0,0 +1,81 @@
+
+1. Running Helm v2 and v3 concurrently on the same cluster:
+   - v2 and v3 release (history) storage are independent of each other. The changes includes the Kubernetes resource for storage and the release object metadata contained in the resource. Releases will also be on a per user namespace instead of using the the Tiller namespace (for example, v2 default Tiller namespace kube-system).v2 uses "ConfigMaps" or "Secrets" under Tiller namespace and `TILLER ...
+   - The only issue could be if Kubernetes cluster scoped resources (e.g. `clusterroles.rbac`) are defined in a chart. The v3 deployment would then fail even if unique in the namepsace as the resources would clash.
```
```suggestion
   - The only issue could be if Kubernetes cluster scoped resources (e.g. `clusterroles.rbac`) are defined in a chart. The v3 deployment would then fail even if unique in the namespace as the resources would clash.
```

hickeyma replied:
done

hickeyma committed:
Update following migration plugin delivery

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Thanks for review comments @php-coder. Updated now.

rimusz commented:
LGTM

hickeyma commented:
Closing this PR as I have created PR https://github.com/helm/helm-www/pull/287 in helm-www where docs reside now.
